### PR TITLE
Predicate matchers and missing Predicate tests. #121

### PR DESCRIPTION
**Pull request by lemonboston**
This pull request adds missing `Predicate` `Matchers` and rewrites tests to use them. Also adds 2 missing unit tests for 'predicate' package classes.
I haven't created unit test for `BinaryPredicate` since it is covered by all the other classes that use it, but let me know if I should. Same for `DelegatingPredicate'.

The matchers have been moved into `PredicateMatcher` as inner classes. It's quite big as a result, but it still seems better than to 'pollute' the package with all these classes which are all predicate matcher related. `PredicateMatcher` can this way expose the necessary convenience methods from a single place. Let me know if you'd prefer instead to have those inner classes as top levels.

Just noticed now that there is hamcrest abstract matcher called `FeatureMatcher`, it could have possibly been used, but I didn't want to spend time with that now. Good to know about that at least for next time.

### CODE REVIEW

**Comment by dmfs** [comment]
Usage of the matchers looks good. Why did you make all the matchers inner classes?

**Comment by lemonboston** [comment]
I've simplified the matchers significantly using `FeatureMatcher` - it turns out to be a very useful base class.
I've also remove not used factory methods, javadocs from the inner classes, so `PredicateMatcher` number of lines have been halved.
I made the matchers inner classes because they will not be needed outside `PredicateMatcher` and I didn't want to pollute the namespace with long named classes unnecessarily*. `PredicateMather` offers the factory methods at a single place, so the matchers are sort of implementation detail to that. With this latest simplification, the class is not that big either.

*I actually added a `PredicateArgumentMatcher` where 'Argument' means the mockito matcher, so we already have 2 classes. It could be confusing I think to add 3-4 more.

Ready for review again. Hopefully the coverage increases as I added a unit test for `PredicateMatcher`.

File: contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ * @author Gabor Keszthelyi
+ */
+public final class PredicateMatcherTest
+{
+
+    @Test
+    public void testSelection()
+    {
+        // Without TC
+        selection("a").matches(
```
Comment (dmfs) [inline contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java]: shouldn't there be any assertion on this line?

File: contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                when(mock(Predicate.class).selection(any(TransactionContext.class))).thenReturn("a").getMock());
+
+        // With TC
+        TransactionContext dummyTc = dummy(TransactionContext.class);
+
+        assertTrue(selection(dummyTc, "a").matches(
```
Comment (dmfs) [inline contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java]: also do the negative check, i.e. is must return `false` if the arguments don't match.

**Review by dmfs — CHANGES_REQUESTED** [review]
Apart from the two minor issues it looks good to me.
I'm still not a big fan of the large number of inner classes. but you can leave it like that.

File: contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java (hunk 1 above)
Comment (lemonboston) [reply contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java]: Right, I missed that, added it.

File: contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java (hunk 2 above)
Comment (lemonboston) [reply contentpal-testing/src/test/java/org/dmfs/android/contentpal/testing/predicates/PredicateMatcherTest.java]: It occured to me, too, afterwards... added some assertions for `false` case as well.

**Comment by lemonboston** [comment]
Ready for check again.
About the inner classes, I totally agree it's usually not good practice, here this 'unit' feels more cohesive this way for me.

**Review by dmfs — APPROVED** [review]
Look good.

**Pull request merged**